yasa/hypno.py

```diff
@@ -48,6 +24,826 @@
+     'REM': 37.5,
+     '%N1': 16.5072,
+     '%N2': 59.0909,
+     '%N3': 15.4306,
+     '%REM': 8.9713}
+    """
+
+    def __init__(self, values, n_stages=5, *, freq="30s", start=None, scorer=None):
+        assert isinstance(values, (list, np.ndarray, pd.Series))
+        assert isinstance(n_stages, int)
+        assert n_stages in [2, 3, 4, 5]
+        assert isinstance(freq, str)
+        assert isinstance(start, (type(None), str, pd.Timestamp))
+        assert isinstance(scorer, (type(None), str, int))
+        if n_stages == 2:
+            accepted = ["S", "W", "SLEEP", "WAKE", "ART", "UNS"]
+            mapping = {"WAKE": 0, "SLEEP": 1, "ART": -1, "UNS": -2}
+        elif n_stages == 3:
+            accepted = ["WAKE", "W", "NREM", "REM", "R", "ART", "UNS"]
+            mapping = {"WAKE": 0, "NREM": 2, "REM": 4, "ART": -1, "UNS": -2}
+        elif n_stages == 4:
+            accepted = ["WAKE", "W", "LIGHT", "DEEP", "REM", "R", "ART", "UNS"]
+            mapping = {"WAKE": 0, "LIGHT": 2, "DEEP": 3, "REM": 4, "ART": -1, "UNS": -2}
+        else:
+            accepted = ["WAKE", "W", "N1", "N2", "N3", "REM", "R", "ART", "UNS"]
+            mapping = {"WAKE": 0, "N1": 1, "N2": 2, "N3": 3, "REM": 4, "ART": -1, "UNS": -2}
+        assert all([val.upper() in accepted for val in values]), (
+            f"{np.unique(values)} do not match the accepted values for a {n_stages} stages "
+            f"hypnogram: {accepted}"
+        )
+        if isinstance(values, pd.Series):
+            # Make sure to remove index if the input is a pandas.Series
+            values = values.to_numpy(copy=True)
+        hypno = pd.Series(values).str.upper()
+        if scorer is None:
+            hypno.name = "Stage"
+        else:
+            hypno.name = scorer
+        # Combine accepted values
+        map_accepted = {"S": "SLEEP", "W": "WAKE", "R": "REM"}
+        hypno = hypno.replace(map_accepted)
+        labels = pd.Series(accepted).replace(map_accepted).unique().tolist()
+        if start is not None:
+            hypno.index = pd.date_range(start=start, freq=freq, periods=hypno.shape[0])
+            timedelta = hypno.index - hypno.index[0]
+        else:
+            fake_dt = pd.date_range(start="2022-12-03 00:00:00", freq=freq, periods=hypno.shape[0])
+            timedelta = fake_dt - fake_dt[0]
+        self._hypno = hypno
+        self._n_epochs = hypno.shape[0]
+        self._freq = freq
+        self._sampling_frequency = 1 / pd.Timedelta(freq).total_seconds()
+        self._timedelta = timedelta
+        self._start = start
+        self._n_stages = n_stages
+        self._labels = labels
+        self._mapping = mapping
+        self._scorer = scorer
+
+    def __repr__(self):
+        return f"{self.hypno}"
+
+    def __str__(self):
+        return f"{self.hypno}"
+
+    @property
+    def hypno(self):
+        """The hypnogram values, stored in a :py:class:`pandas.Series`."""
+        # Q: Should this be called `hyp.stage`?
```

Comment:
Let me offer a clarification here, because I think I misunderstood what was going on here at first. I've been playing around with this a bit since. I think it would be nice to have it set up like this:

* `Hypnogram.hypno` = The pandas.Series
* `Hypnogram.values` = numpy array of stages as integers
* `Hypnogram.names` = numpy array of stages as strings ("names" or something similar)
* `Hypnogram.stages` = list of possible labels (currently `.labels`)
* `Hypnogram.mapping_str2int` and `Hypnogram.mapping_int2str` for the two respective mapping dictionaries.

This would replace `Hypnogram.as_int()` and `.mapping`. The idea would be that there is no "as something" because there is no assumption that the hypnogram is normally one way and then converted to another. It's like the Hypnogram is an all-encompassing thing that contains lots of info (integer format, string format, scorer etc.) and you can pull anything out directly.

Thoughts?

Reply:
This is quite a major change so let me think about it. A few first comments:

- Currently, users can define a custom mapping with `hyp.mapping = {}`, which is a nice feature. How would you implement that with this new structure?
- I am worried about having too many class attributes — the simpler, the better. What about `values` for strings and `values_int` for integers?
- I can rename `labels` to `stages` if that makes more sense

Reply:
Don't worry about this, leave it all as-is (including `labels`). I think I spoke too quickly, the current setup works great. For any advanced user, working with the `Hypnogram.hypno` Series solves any need for creativity. I've been working on re-implementing the Evaluation PR the last few hours and honestly this new class is _sooo_ nice.